<a href="https://colab.research.google.com/github/uyangas/Visualizations-in-Python/blob/main/AirBnB_new_user_booking_destination_plotly.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# AirBnB new user booking destination (Plotly)

https://www.kaggle.com/c/airbnb-recruiting-new-user-bookings/data


`train_users.csv` - the training set of users
- `id`: user id
- `date_account_created`: the date of account creation
- `timestamp_first_active`: timestamp of the first activity, note that it can be earlier than date_account_created or date_first_booking because a user can search before signing up
- `date_first_booking`: date of first booking
- `gender`
- `age`
- `signup_method`
- `signup_flow`: the page a user came to signup up from
- `language`: international language preference
- `affiliate_channel`: what kind of paid marketing
- `affiliate_provider`: where the marketing is e.g. google, craigslist, other
- `first_affiliate_tracked`: whats the first marketing the user interacted with before the signing up
- `signup_app`
- `first_device_type`
- `first_browser`

`country_destination`: this is the target variable you are to predict

`sessions.csv` - web sessions log for users

`user_id`: to be joined with the column 'id' in users table
- `action`
- `action_type`
- `action_detail`
- `device_type`
- `secs_elapsed`

`countries.csv` - summary statistics of destination countries in this dataset and their locations

`age_gender_bkts.csv` - summary statistics of users' age group, gender, country of destination
sample_submission.csv - correct format for submitting your predictions


In [ ]:
from google.colab import files
files.upload()

In [2]:
# create "kaggle" directory
!mkdir ~/.kaggle
# save kaggle.json to "kaggle" directory
! cp kaggle.json ~/.kaggle/
# change the permission of the file
! chmod 600 ~/.kaggle/kaggle.json

In [3]:
# download the dataset
! kaggle competitions download -c airbnb-recruiting-new-user-bookings --force

  0% 0.00/2.46k [00:00<?, ?B/s]
100% 2.46k/2.46k [00:00<00:00, 5.05MB/s]
 96% 57.0M/59.1M [00:00<00:00, 76.0MB/s]
100% 59.1M/59.1M [00:00<00:00, 133MB/s] 
  0% 0.00/546 [00:00<?, ?B/s]
100% 546/546 [00:00<00:00, 528kB/s]
  0% 0.00/1.03M [00:00<?, ?B/s]
100% 1.03M/1.03M [00:00<00:00, 157MB/s]
  0% 0.00/478k [00:00<?, ?B/s]
100% 478k/478k [00:00<00:00, 118MB/s]
  0% 0.00/4.07M [00:00<?, ?B/s]
100% 4.07M/4.07M [00:00<00:00, 66.1MB/s]


In [4]:
# create directory to unzip the data
! mkdir ~/.airbnb_booking

In [5]:
# unzip the files
! unzip countries.csv.zip -d airbnb_booking
! unzip train_users_2.csv.zip -d airbnb_booking
! unzip age_gender_bkts.csv.zip -d airbnb_booking
! unzip sessions.csv.zip -d airbnb_booking
! unzip test_users.csv.zip -d airbnb_booking

Archive:  countries.csv.zip
  inflating: airbnb_booking/countries.csv  
Archive:  train_users_2.csv.zip
  inflating: airbnb_booking/train_users_2.csv  
Archive:  age_gender_bkts.csv.zip
  inflating: airbnb_booking/age_gender_bkts.csv  
Archive:  sessions.csv.zip
  inflating: airbnb_booking/sessions.csv  
Archive:  test_users.csv.zip
  inflating: airbnb_booking/test_users.csv  


In [7]:
! pip install impyute

## 1. Preprocessing

In [22]:
# import necessary packages
import pandas as pd
import numpy as np
import plotly.io as pio
import plotly.express as px
import plotly.graph_objects as go
import plotly.figure_factory as ff
from plotly.subplots import make_subplots
from datetime import datetime
from impyute.imputation.cs import mice
from sklearn.preprocessing import LabelEncoder
import pickle

In [9]:
# load the datasets
countries = pd.read_csv("airbnb_booking/countries.csv")
train = pd.read_csv("airbnb_booking/train_users_2.csv")
test = pd.read_csv("airbnb_booking/test_users.csv")
age_gender = pd.read_csv("airbnb_booking/age_gender_bkts.csv")
sessions = pd.read_csv("airbnb_booking/sessions.csv")

In [10]:
# merge train and test datasets
test['country_destination'] = 'Test'
all = pd.concat([train, test], axis=0, ignore_index=True)

In [11]:
all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275547 entries, 0 to 275546
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   id                       275547 non-null  object 
 1   date_account_created     275547 non-null  object 
 2   timestamp_first_active   275547 non-null  int64  
 3   date_first_booking       88908 non-null   object 
 4   gender                   275547 non-null  object 
 5   age                      158681 non-null  float64
 6   signup_method            275547 non-null  object 
 7   signup_flow              275547 non-null  int64  
 8   language                 275547 non-null  object 
 9   affiliate_channel        275547 non-null  object 
 10  affiliate_provider       275547 non-null  object 
 11  first_affiliate_tracked  269462 non-null  object 
 12  signup_app               275547 non-null  object 
 13  first_device_type        275547 non-null  object 
 14  firs

In [12]:
# convert datetime
all['date_account_created'] = pd.to_datetime(all['date_account_created'])
all['date_first_booking'] = pd.to_datetime(all['date_first_booking'])
all['timestamp_first_active'] = pd.to_datetime(all['timestamp_first_active'], format='%Y%m%d%H%M%S')

In [13]:
print(f"Total number of users: {all.shape[0]}, Number of variables: {all.shape[1]}")

Total number of users: 275547, Number of variables: 16


In [14]:
all.head()

id date_account_created  ... first_browser country_destination
0  gxn3p5htnn           2010-06-28  ...        Chrome                 NDF
1  820tgsjxq7           2011-05-25  ...        Chrome                 NDF
2  4ft3gnwmtx           2010-09-28  ...            IE                  US
3  bjjt8pjhuk           2011-12-05  ...       Firefox               other
4  87mebub9p4           2010-09-14  ...        Chrome                  US

[5 rows x 16 columns]

In [119]:
for col in ['gender','signup_method','signup_flow','language','affiliate_channel','affiliate_provider','signup_app','first_device_type','first_browser','country_destination']:
  all[col] = all[col].astype('category')

### 1.1. Handling missing values

In [23]:
# plot the percentage of NA values in each column
fig = px.bar(x='Columns', 
             y='value', 
             data_frame=(all.isna().sum()/all.shape[0])\
             .reset_index()\
             .assign(Non_NA = lambda x: 1-x[0])\
             .rename({'index':'Columns',0:'NA_values'},axis=1)\
             .melt('Columns',['NA_values', 'Non_NA']),
             color='variable',
             orientation='v',
             category_orders={"Columns": ["NA values", "Non NA values"]},
             color_discrete_sequence=["rgb(0,184,184)", "rgb(227,227,227)"],
             hover_name='Columns',
             title='Percentage of NA values')

fig.update_layout(width=1000, height=400,
                  yaxis={'title':'Percentage'},
                  xaxis={'title':''},
                  plot_bgcolor='#fff')
fig.show()

`date_first_booking`, `age` and `first_affiliate_tracked` columns contained `67.7%`,  `42.4%` and `2.2%` NA values respectively.

In [16]:
# explore NA values of date_first_booking
all[all.date_first_booking.isna()][['id','date_account_created','timestamp_first_active','date_first_booking','country_destination']].head(10)

id date_account_created  ... date_first_booking country_destination
0   gxn3p5htnn           2010-06-28  ...                NaT                 NDF
1   820tgsjxq7           2011-05-25  ...                NaT                 NDF
11  om1ss59ys8           2010-01-05  ...                NaT                 NDF
13  dy3rgx56cu           2010-01-05  ...                NaT                 NDF
14  ju3h98ch3w           2010-01-07  ...                NaT                 NDF
16  2dwbwkx056           2010-01-07  ...                NaT                 NDF
18  cxlg85pg1r           2010-01-08  ...                NaT                 NDF
23  jha93x042q           2010-01-11  ...                NaT                 NDF
24  7i49vnuav6           2010-01-11  ...                NaT                 NDF
26  bjg0m5otl3           2010-01-12  ...                NaT                 NDF

[10 rows x 5 columns]

In [17]:
# count the destination countries
all[all.date_first_booking.isna()]['country_destination'].value_counts()

NDF     124543
Test     62096
Name: country_destination, dtype: int64

To explore more about NA values of `date_first_booking` column. When `date_first_booking` is null, the person didn't place any booking. It seems it is not necessary to impute this value because imputation would alter the meaning of this variable.

Let's look at `age` column.

In [18]:
# descriptive statistics of age
all.age.describe()

count    158681.000000
mean         47.145310
std         142.629468
min           1.000000
25%          28.000000
50%          33.000000
75%          42.000000
max        2014.000000
Name: age, dtype: float64

The maximum value that the `age` column contains is 2014 which could be a typo or an error because it is unlikely that people can live up to over 2000 years, yet alone over 100 years. The minimum value of `age` is 1 which is highly unlikely as well.

In [19]:
# age greater than 100
all['age'][all.age>100].value_counts().head(10)

105.0     1351
2014.0     710
110.0      228
104.0       52
101.0       40
102.0       39
2013.0      39
109.0       36
103.0       30
107.0       28
Name: age, dtype: int64

In [84]:
pio.templates.default = "ggplot2"
# create subplots
fig = make_subplots(1, 2, 
                    subplot_titles=["Histogram of Age", 'Boxplot of Age'],
                    column_widths=[0.3, 0.3])
# add histogram trace to the subplot
fig.add_trace(go.Histogram(x=all[all.age<110].age, 
                           marker_color='rgb(0,186,202)'), 
              1,1)
# add boxplot trace to the subplot
fig.add_trace(go.Box(x=all[all.age<110].age, marker_color='rgb(0,189,166)'), 1,2)
# update the figure size
fig.update_layout(width=1200, height=400, showlegend=False)
fig.show()

From the plots above, `age` has several values less than 10 which could most likely errors. Therefore, extreme values i.e. `age` less than 10 and greater than 100 are considered as outliers, removed and filled with imputation.

In [49]:
# remove extreme values and add np.nan
all['age'][(all['age'] <10)|(all['age'] >100)] = np.nan

Since `age` is continuous variable and filling `NA` with mean or median value would cause distribution change, `MICE` method was used. To use `MICE` all categorical values must be numerical. Therefore, the values of caterical attributes were encoded using `LabelEncoder`.

In [50]:
# create a copy of all dataframe
all_transformed = all.copy()
# encode the values of each categorical attribute
for col in ['gender','signup_method','signup_flow','language','affiliate_channel',
            'affiliate_provider','signup_app','first_device_type','country_destination','first_browser']:
  # call the object
  le = LabelEncoder()
  # fit to the column
  le.fit(all_transformed[col])
  # transform the values
  all_transformed[col] = le.transform(all_transformed[col])

In [51]:
all_transformed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275547 entries, 0 to 275546
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   id                       275547 non-null  object        
 1   date_account_created     275547 non-null  datetime64[ns]
 2   timestamp_first_active   275547 non-null  datetime64[ns]
 3   date_first_booking       88908 non-null   datetime64[ns]
 4   gender                   275547 non-null  int64         
 5   age                      155932 non-null  float64       
 6   signup_method            275547 non-null  int64         
 7   signup_flow              275547 non-null  int64         
 8   language                 275547 non-null  int64         
 9   affiliate_channel        275547 non-null  int64         
 10  affiliate_provider       275547 non-null  int64         
 11  first_affiliate_tracked  269462 non-null  category      
 12  signup_app      

In [52]:
# use MICE to impute age variable
imputed = mice(all_transformed.drop(['id',
                                     'country_destination',
                                     'date_first_booking',
                                     'date_account_created',
                                     'timestamp_first_active',
                                     'first_affiliate_tracked'],axis=1).values)
imputed[:,1]

array([33.26362774, 38.        , 56.        , ..., 37.00965388,
       38.57737665, 49.        ])

In [56]:
# save imputed age variable as pickle 
with open('age.pkl','wb') as f:
  pickle.dump(imputed[:,1], f)

In [58]:
# replace NA values with imputed values
all['age'] = imputed[:,1].astype('int')
# delete unnecessary dataframes
del imputed, all_transformed

Now, let's look at `first_affiliate_tracked` column.

In [64]:
print(f"Unique values: {all['first_affiliate_tracked'].unique()}")
print(f"Number of NA values: {all['first_affiliate_tracked'].isna().sum()}")

Unique values: ['untracked', 'omg', NaN, 'linked', 'tracked-other', 'product', 'marketing', 'local ops']
Categories (7, object): ['untracked', 'omg', 'linked', 'tracked-other', 'product', 'marketing',
                         'local ops']
Number of NA values: 6085


In [116]:
# dataframe with value counts
fat_count = all[['id','first_affiliate_tracked']].groupby('first_affiliate_tracked').count().reset_index()

# pie chart
fig = go.Figure()
colors = ['rgb(203,203,203)',
          'rgb(192,192,192)',
          'rgb(120,120,120)',
          'rgb(0,186,202)',
          'rgb(169,169,169)',
          'rgb(140,140,140)',
          'rgb(130,130,130)',
          'rgb(226,226,226)']
fig.add_trace(go.Pie(labels=fat_count['first_affiliate_tracked'],
                     values=fat_count['id'],
                     hole=.3,
                     rotation=260,
                     textinfo='label+percent',
                     showlegend=False,
             ))
fig.update_layout(width=450, height=450, title="Precentage of values in first_affiliate_tracked")
fig.update_traces(marker=dict(colors=colors))
fig.show()

`first_affiliate_tracked` column has only 6,085 `NA` values i.e. 2.21% of the total number of rows. Based on the number of categories and the number of values in each category, `NA` values were treated as a separate category and filled as `No Value`.

In [72]:
# filled missing values with No Value instead of other types of filling methods
all['first_affiliate_tracked'] = all['first_affiliate_tracked'].astype('str').fillna('No Value').astype('category')

In [120]:
all.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 275547 entries, 0 to 275546
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   id                       275547 non-null  object        
 1   date_account_created     275547 non-null  datetime64[ns]
 2   timestamp_first_active   275547 non-null  datetime64[ns]
 3   date_first_booking       88908 non-null   datetime64[ns]
 4   gender                   275547 non-null  category      
 5   age                      275547 non-null  int64         
 6   signup_method            275547 non-null  category      
 7   signup_flow              275547 non-null  category      
 8   language                 275547 non-null  category      
 9   affiliate_channel        275547 non-null  category      
 10  affiliate_provider       275547 non-null  category      
 11  first_affiliate_tracked  275547 non-null  category      
 12  signup_app      